src/Agent.Worker/Build/TrackingManager.cs

```diff
@@ -137,6 +144,112 @@ public void UpdateJobRunProperties(IExecutionContext executionContext, TrackingC
             WriteToFile(file, config);
         }
 
+        public void DiscoverExpiredGarbage(IExecutionContext executionContext, TimeSpan expiration)
+        {
+            Trace.Entering();
+            Trace.Info("Scan all SourceFolder tracking files.");
+            string searchRoot = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Work), Constants.Build.Path.SourceRootMappingDirectory);
+            if (!Directory.Exists(searchRoot))
+            {
+                executionContext.Output($"SourceRootMapping directory '{searchRoot}' doesn't exist, skip garbage discovery.");
+                return;
+            }
+
+            var allTrackingFiles = Directory.EnumerateFiles(searchRoot, Constants.Build.Path.TrackingConfigFile, SearchOption.AllDirectories);
+            Trace.Verbose($"Find {allTrackingFiles.Count()} tracking files.");
+
+            executionContext.Output($"Directory expiration limit: {expiration.TotalDays} days.");
+            executionContext.Output($"Current UTC: {DateTime.UtcNow.ToString("o")}");
+
+            // scan all sourcefolder tracking file, find which folder has never been used since UTC-expiration
+            // the scan and garbage discovery should be best effort.
+            // if the tracking file is in old format, just delete the folder since the first time the folder been use we will convert the tracking file to new format.
+            foreach (var trackingFile in allTrackingFiles)
+            {
+                try
+                {
+                    executionContext.Output($"Evaluate BuildDirectory tracking file: {trackingFile}");
+                    TrackingConfigBase tracking = LoadIfExists(executionContext, trackingFile);
+
+                    // detect whether the tracking file is in new format.
+                    TrackingConfig newTracking = tracking as TrackingConfig;
+                    if (newTracking == null)
+                    {
+                        LegacyTrackingConfig legacyConfig = tracking as LegacyTrackingConfig;
+                        ArgUtil.NotNull(legacyConfig, nameof(LegacyTrackingConfig));
+
+                        Trace.Verbose($"{trackingFile} is a old format tracking file.");
+
+                        executionContext.Output($"Mark tracking file '{trackingFile}' for GC, since it never been used.");
+                        MarkForGarbageCollection(executionContext, legacyConfig);
+                        IOUtil.DeleteFile(trackingFile);
```

Comment:
i'm not sure how i feel about deleting all of the legacy tracking files :)

maybe delete is fine. although i'm wondering if we should fallback to timestamp on the file.